### DISCUSSION TITLE
Lists

### DISCUSSION THREAD

**Discussion by jghauser** [nvim-neorg/neorg#321] [General]
First of all, thanks a lot for all the hard work on neorg. I've been looking into it every now and then over the last few months, and it's great to see it so quickly reach a state where I'm considering switching to it fully for my notes and GTD.

One of my uses is writing notes to structure my thinking when preparing articles. This usually involves really long lists of bullet points. In this regard, I've come across some things that are unclear to me.

1. How do ordered and unordered lists combine? For example, I want to add a ordered list as a subitem of an unordered list (or vv). Do the entries in the ordered list need to have an indentation level that is one level higher than the unordered list or does switching from unordered to ordered (and vv) 'reset' the indentation syntax?

Is this correct?

```
~ item one
 -- something
 -- else
~ item two
```

Or this?

```
~ item one
  - something
  - else
~ item two
```

It looks like the concealer works correctly (showing `2. item two`) with the first syntax. Is that correct like this?

2. If the first syntax above is correct, then that means we can only ever have 7 levels of indentation. This sounds like a lot, but in practice I often reach this limit quite quickly. I feel like orgmode/markdown syntax (without repeated modifiers) is much more flexible in this way.
3. Moreover, such a syntax means that an ordered list at the second level will always be `A...Z` rather than `1...n` because it's at the second indentation level even if no ordered list of a lower level exists. This isn't what I would intuitively expect.
4. I often find myself at a very high level of indentation. It's quite tedious having to remember the indentation level and often type 5-7 modifiers on each line. It would be very useful if neorg could automatically create the modifiers when going to a new line (e.g. with `<cr>` or `o`). Is such a thing planned? Switching from ordered to unordered would still be tedious, but at least simply continuing lists would be made much easier.
5. Concealer issues. 
    - The concealer only updates when I actually go to the line containing, e.g., `~ item two` and thus the number might be wrong if I for example add another numbered item after `~ item one`. 
    - the `#ordered` tag doesn't seem to be taken into account.

I just typed this down rather quickly, so if anything is unclear please do ask ;).

**Comment by mrossinek** [comment]
1. Yes, you will need to increase the indentation (i.e. the modifier length) in order to do nesting. Thus, your first example is correct.
2. Yes, currently the indentation levels are capped at 6 (same as HTML headings) although you bringing this up made me notice that there is a bug in our TS parser. Headings allow you to use arbitrarily many `*` and will fall back to interpreting anything longer than 6 chars as heading level 6. The same should be done for lists, quotes, etc. instead of just breaking the parser. I will fix that in a moment! (EDIT: https://github.com/nvim-neorg/tree-sitter-norg/commit/e3d477957630aa0786358bffb28f3d2bbcdf7aa9)
3. This is something we are not likely to change simply because this concealing logic is already complex enough as it is. However, if you are fit with some Treesitter queries and Lua you could change your concealing configuration to take this into account. You would need to check for an ordered list item what it's parent kind is and adapt the concealing icons accordingly.. I'm not sure it's worth the runtime overhead though.
4. An indentation engine is being worked. I assume that this will also include clever list continuation.
5. a) I'm not sure I fully understand the exact issue. The concealing updates are done with `vim.schedule` calls so there could be a delay. But if you have not edited the file the icons should all be there after a short while. b) Yes, this is not yet being taken into account at all. These tags are mostly relevant for format conversions (once that becomes more supported) as the concealing was more an ad-hoc idea to show off what Neovim is capable off.

**Comment by jghauser** [comment]
Sorry for the slow reply and thanks for the answer! I just have a small comment about 5.a which I think might be a bug. If I have a list like so:
![20220215_22h57m48s_grim](https://user-images.githubusercontent.com/10319377/154156461-35096440-5c32-4a67-9929-12b0360e9a79.jpeg)
and I add a new entry between the first two, the numbering for the last two entries doesn't update:
![20220215_22h58m04s_grim](https://user-images.githubusercontent.com/10319377/154156638-9cb8717a-d20b-4656-8fc7-0475e89d34a4.jpeg)
The update only happens after I move to those lines and enter and leave insert mode.
